### ISSUE TITLE
[1.8.0-rc1] Cannot kill or stop daemon containers before a reboot

### ISSUE THREAD

**Issue by jgkamat** [docker/docker#15101]
Description of problem:
Cannot kill or stop daemon containers on the current docker experimental

`docker version`:
```
Client:
 Version:      1.8.0-rc1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   4f1c66a
 Built:        Sat Jul 25 05:54:28 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.0-rc1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   4f1c66a
 Built:        Sat Jul 25 05:54:28 UTC 2015
 OS/Arch:      linux/amd64
```

`docker info`:
```
Containers: 2
Images: 3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 7
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-41-generic
Operating System: Ubuntu 14.04 LTS
CPUs: 1
Total Memory: 3.66 GiB
Name: mechjeb
ID: 5OBA:PPB3:SK7K:YVNU:Y5KR:E36L:Q55E:G66W:F7Z5:UO6D:JNLO:ZQQY
WARNING: No swap limit support
```

`uname -a`:
```
Linux mechjeb 3.16.0-41-generic #57-Ubuntu SMP Thu Jun 18 08:44:16 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

Environment details (AWS, VirtualBox, physical, etc.):
Running on Ubuntu 14.04 on AWS

How reproducible:
Always

Steps to Reproduce:
```
root@mechjeb:~# wget -qO- https://experimental.docker.com/ | sh
<<Install successfull>>
root@mechjeb:~# docker run -d busybox top
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
cf2616975b4a: Pull complete 
6ce2e90b0bc7: Pull complete 
8c2e06607696: Already exists 
Digest: sha256:38a203e1986cf79639cfb9b2e1d6e773de84002feea2d4eb006b52004ee8502d
Status: Downloaded newer image for busybox:latest
67d23cabe882a20a93f4bf08d84347f240cf64886764507b935a17ca9b3a1f47
root@mechjeb:~# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
67d23cabe882        busybox             "top"               3 seconds ago       Up 3 seconds                            lonely_tesla
root@mechjeb:~# docker stop 67d23
Error response from daemon: Cannot stop container 67d23: permission denied
Error: failed to stop containers: [67d23]
root@mechjeb:~# docker kill 67d23
Error response from daemon: Cannot kill container 67d23: permission denied
Error: failed to kill containers: [67d23]
root@mechjeb:~# docker rm -f 67d23
Error response from daemon: Cannot destroy container 67d23: Could not kill running container, cannot remove - permission denied
Error: failed to remove containers: [67d23]
root@mechjeb:~#
```

Actual Results:
Container not removed/killed/stopped (Until a reboot is issued)

Expected Results:
Container killed/removed/stopped

**Comment by coolljt0725** [comment]
I have met the same problem, but I can't reproduce after reboot my system

**Comment by jgkamat** [comment]
hmm, on many CI machines docker is installed on a clean machine and the machine should not have to be rebooted (unless a kernel upgrade is needed).

**Comment by Evalle** [comment]
I also can reproduce this issue on my host and it's gone after reboot, but I agreed with @jgkamat: this workaround is not a good choice.

**Comment by mountkin** [comment]
Perhaps the install script didn't reload the apparmor service after installation.
I installed docker with the script from `https://get.docker.com` but the daemon failed to start.
I found "Error starting daemon: error initializing graphdriver: permission denied" in `/var/log/upstart/docker.log`
After running `service apparmor reload` the daemon started as expected.
I use a custom graph path (/root/docker).

```
Linux local-dev 3.13.0-24-generic #47-Ubuntu SMP Fri May 2 23:30:00 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jgkamat** [comment]
a `service apparmor reload` does not fix this issue. (This issue goes away from a reboot though). This should probably be fixed before 1.8 is released, this is present in 1.8.0_RC1 as well.

**Comment by chenchun** [comment]
I've also seeing this with kernel log.
```
Jul 31 13:59:55 chenchun-T450 kernel: [253415.594750] audit: type=1400 audit(1438322395.835:317): apparmor="DENIED" operation="signal" profile="docker-default" pid=3060 comm="docker" requested_mask="receive" denied_mask="receive" signal=kill peer="/usr/bin/docker"
```
Kernel log seems to tell its `3060` who tried to send a kill signal. And `3060` is a thread of docker daemon process.
```
root@chenchun-T450:/home/chenchun/project/go/src/github.com/docker/docker# docker kill $(docker ps -q)
Error response from daemon: Cannot kill container 640ce36f933c: permission denied
FATA[0000] Error: failed to kill one or more containers 
root@chenchun-T450:/home/chenchun/project/go/src/github.com/docker/docker# ps -ef | grep 3060
root      3431 31564  0 14:06 pts/19   00:00:00 grep --color=auto 3060
root@chenchun-T450:/home/chenchun/project/go/src/github.com/docker/docker# cat /var/run/docker.pid 
3015root@chenchun-T450:/home/chenchun/project/go/src/github.com/docker/docker# 
root@chenchun-T450:/home/chenchun/project/go/src/github.com/docker/docker# cat /proc/3060/cmdline 
/usr/bin/docker-d--bip192.168.1.0/16-D-Htcp://0.0.0.0:2375-Hunix:///var/run/docker.sockroot@chenchun-T450:/home/chenchun/project/go/src/github.com/docker/docker# 
root@chenchun-T450:/proc/3060# ps -ef -T | grep 3060
root      3015  3060     1  0 13:54 ?        00:00:00 /usr/bin/docker -d --bip 192.168.1.0/16 -D -H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock

```

**Comment by BenHall** [comment]
I'm also experiencing this since upgrading to RC2. 

```
Error response from daemon: Cannot destroy container st_ben_hall_sHMEgtoUNSBhPCyZAAAU: Could not kill running container, cannot remove - permission denied
Error: failed to remove containers: [st_ben_hall_sHMEgtoUNSBhPCyZAAAU]
```

Rebooting daemon/server stops the container and let's me delete it. However the problem reoccurs. 

I tried rollbacking back to 1.6.3 but was still getting the same error. 

Output from syslog

```
Aug  7 15:42:33 ns3001358 docker[1215]: time="2015-08-07T15:42:33.878809301+02:00" level=error msg="Handler for DELETE /containers/{name:.*} returned error: Cannot destroy container st_ben_hall_sHMEgtoUNSBhPCyZAAAU: Could not kill running container, cannot remove - permission denied"
Aug  7 15:42:33 ns3001358 docker[1215]: time="2015-08-07T15:42:33.878904948+02:00" level=error msg="HTTP Error" err="Cannot destroy container st_ben_hall_sHMEgtoUNSBhPCyZAAAU: Could not kill running container, cannot remove - permission denied" statusCode=500
```


Stopping the daemon results in

```
Aug  7 15:43:25 ns3001358 systemd[1]: Stopping Docker Application Container Engine...
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.797114245+02:00" level=info msg="Processing signal 'terminated'"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.802832518+02:00" level=info msg="Failed to send SIGTERM to the process, force killing"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.803062531+02:00" level=error msg="Stop container cfa1f7f36703c26da2f819cd46edea671471f3b6fa6bff00b070b71179de3ff5 with error: permission denied"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.803615548+02:00" level=info msg="Failed to send SIGTERM to the process, force killing"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.803848503+02:00" level=error msg="Stop container 0020c38ad8be867568edb96323a6fa45d6ca7ba43fc29da71e9ff4d3f99391dd with error: permission denied"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.804146793+02:00" level=info msg="Failed to send SIGTERM to the process, force killing"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.804397579+02:00" level=error msg="Stop container 0cea734cec868badd739386f27386eece1f45735671a61cd28b5097e0832695e with error: permission denied"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.804628142+02:00" level=info msg="Failed to send SIGTERM to the process, force killing"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.804816348+02:00" level=error msg="Stop container 241745c6d9bbec3b846f4eec121c6f1d438df2f6cd8acabc5c3233e5e546c38b with error: permission denied"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.805015624+02:00" level=info msg="Failed to send SIGTERM to the process, force killing"
Aug  7 15:43:25 ns3001358 docker[1215]: time="2015-08-07T15:43:25.805157980+02:00" level=error msg="Stop container 912295cad2bdaee9a26e4299ced4ef81b580914da208ae4b0a7e3c8aace8c9a2 with error: permission denied"
Aug  7 15:43:35 ns3001358 docker[1215]: time="2015-08-07T15:43:35.803605727+02:00" level=info msg="Container 4859bed294c77fcb652f966c1f2f501c998a672c7b6997a00a2673ae21c6d060 failed to exit within 10 seconds of SIGTERM - using the force"
```

I turned on daemon debug mode but didn't show much more.

```
Aug  7 15:55:31 ns3001358 docker[6577]: time="2015-08-07T15:55:31.327974247+02:00" level=debug msg="Calling POST /containers/{name:.*}/stop"
Aug  7 15:55:31 ns3001358 docker[6577]: time="2015-08-07T15:55:31.328049304+02:00" level=info msg="POST /v1.20/containers/st_ben_hall_3E705DcjB0efCoNAAAX/stop?t=10"
Aug  7 15:55:31 ns3001358 docker[6577]: time="2015-08-07T15:55:31.328540762+02:00" level=debug msg="Sending 15 to 0bf4b4729eab25bddb5fc31cb2dd1a937bc5b33d923547073c17cfa48b4a47fc"
Aug  7 15:55:31 ns3001358 docker[6577]: time="2015-08-07T15:55:31.328823156+02:00" level=info msg="Failed to send SIGTERM to the process, force killing"
Aug  7 15:55:31 ns3001358 docker[6577]: time="2015-08-07T15:55:31.328863276+02:00" level=debug msg="Sending 9 to 0bf4b4729eab25bddb5fc31cb2dd1a937bc5b33d923547073c17cfa48b4a47fc"
```

**Comment by BenHall** [comment]
A `kill -9` on the PID of the container will then allow you to delete it.

**Comment by jgkamat** [comment]
This problem is no longer an issue for me on the latest experimental and 1.8 rc3. I'm going to close this issue unless others are still seeing this.

**Comment by BenHall** [comment]
I'm still seeing this.

```
# docker version
Client:
 Version:      1.8.0-rc3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   4ce5440
 Built:        Fri Aug  7 13:01:29 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.0-rc3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   4ce5440
 Built:        Fri Aug  7 13:01:29 UTC 2015
 OS/Arch:      linux/amd64
```

**Comment by thaJeztah** [comment]
ping @calavera in case this needs additional research before the release

**Comment by BenHall** [comment]
Just as an update on this. I sent a number of hours investigating, looking in the docker code but couldn't see anything. This was happening on 15.04, potentially a systemd issue.

1) Changing docker directories didn't resolve issue.
2) Rolling back to 1.6.2 didn't resolve issue.
3) If I launched the daemon from the command line it would be fine. Even when pointed to existing data directory.
4) I could stop the containers with systemctl directly.
5) If I launched a container using `docker -d -t ubuntu bash` it was unable to be stopped. Other containers running non-bash commands could be stopped. I think this was something to do with how the underlying process handled SIGKILL commands, it was the same with a Golang process too.

Sadly I've since destroyed the instance as being unable to remove containers had knock-on effects to the rest of the system.

**Comment by jgkamat** [comment]
Ah, I was testing this on 14.04, so this could be a systemd issue (now?). 
I'll keep trying to reproduce it and see if I can get any more information.

For me, This issue wasn't present on 1.7.1, and was an issue on 1.8 experimental before rc1.
I noticed symptom (3) as well.
I tried (5) with many commands, and I couldn't find any commands which worked. Not sure if I tried enough though.

As of now, I cannot reproduce this anymore (ubuntu 14.04 on 1.8 rc3). I'll let you know if I'm able to reproduce it on 15.04 once I get some time.

**Comment by BenHall** [comment]
Just tried on a clean digital ocean droplet and the above command worked with RC3 and the container could be deleted...

**Comment by calavera** [comment]
There were several changes in the way we handle appArmor policies merged in RC1 and RC2 that were completely reverted in RC3. If you are still experiencing this issue, it's probably caused because you still have a profile under `/etc/apparmor`.

Sorry for the inconvenience and thanks for reporting this, it makes it even more clear that the decision of reverting those changes was right.

I'm going to close this issue :confetti_ball:

**Comment by JohnyDays** [comment]
I still have this issue on docker 1.8.1. Removing `/etc/apparmor.d/docker` and running `sudo service apparmor reload` appears to fix it

**Comment by JohnyDays** [comment]
This problem has made any version of docker unusable on my development machine. Having tried 1.8.1 1.8.0 1.9.0 (experimental) and 1.7.1, all of them exhibit the same problem, making it impossible to delete containers. I have also tried disable AppArmor which did not fix it either. Is there some way to fully reset anything docker related and start from scratch? I have tried deleting /etc/docker and /var/lib/docker along with apt remove and purge, but the issue persists.

**Comment by jgkamat** [comment]
@JohnyDays if this issue persists even after you reboot your machine, you might be seeing a different issue altogether! I didn't see this issue until late in the 1.8.0 experimental phase. Are you seeing this issue when you manually run the daemon with `docker daemon`? I don't remember having this issue when starting the daemon manually...

Maybe you and @BenHall are both seeing the same issue, as you are both on 15.04? (I've only been able to test on 14.04, sorry!)

**Comment by JohnyDays** [comment]
I have rebooted my machine several times indeed, and the issue persists. I have also uninstalled and reinstalled several versions of docker, rebooting along the way. Over those iterations, after installing I can create containers, but not delete them, then I remove /etc//apparmor/docker, and I can delete them but not create/start them. I have now begun reinstalling ubuntu on my machine :cry: sorry if that doesn't help solve the issue

**Comment by jacohend** [comment]
Same issue here with 1.90 experimental. Unusable for development.

**Comment by jgkamat** [comment]
Hmm, everyone seems to be having this problem on 15.04, so it might be something with systemd? I can't reopen this issue for some reason though...

**Comment by thaJeztah** [comment]
I'll reopen this issue, so that we can further investigate, also;

ping @ewindisch could you have a look here, because you've been working on the AppArmor changes

**Comment by JohnyDays** [comment]
It might be worth mentioning that after reinstalling Ubuntu, I haven't had this problem reappear. I'm currently running docker 1.9 and Ubuntu 15.04

**Comment by jacohend** [comment]
@JohnyDays I hear you, but I think the point is more that people shouldn't have to reinstall their host systems in order to upgrade docker.

**Comment by diogomonica** [comment]
Same thing just happened to me on Ubuntu 15:04:
```
Linux ubuntu 4.2.0-16-generic #19-Ubuntu SMP Thu Oct 8 15:35:06 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

```
Containers: 3
Images: 3
Server Version: 1.10.0-dev
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-16-generic
Operating System: Ubuntu 15.10
CPUs: 1
Total Memory: 992.9 MiB
Name: dockercon-demo-2
ID: AIMU:53J6:ENMF:HFNN:4EWI:66B7:2ANK:PNWW:HPPY:Y7JN:KMJP:7E6R
WARNING: No swap limit support
Experimental: true
```

```
# docker kill loving_bell
Error response from daemon: Cannot kill container loving_bell: permission denied
Error: failed to kill containers: [loving_bell]
```

**Comment by thaJeztah** [comment]
@diogomonica have you checked your syslog for apparmor messages?

**Comment by diogomonica** [comment]
Yeah, there we go:

```
Nov 16 11:06:23 diogo-dockercon-demo kernel: [ 1059.048784] audit: type=1400 audit(1447689983.403:334): apparmor="DENIED" operation="signal" profile="docker-default" pid=17651 comm="docker" requested_mask="receive" denied_mask="receive" signal=kill peer="/usr/bin/docker"
```

**Comment by jfrazelle** [comment]
this is fixed via https://github.com/docker/docker/pull/18178/files

**Comment by fk0** [comment]
I'm wondering if this issue resolved in 1.11? I'm still having it in this version.
```
$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 34
Server Version: 1.11.0-dev
Storage Driver: aufs
 Root Dir: /mnt/xfs/docker/aufs
 Backing Filesystem: xfs
 Dirs: 128
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 4.2.0-30-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.673 GiB
Name: dell
ID: ZQ5F:RTRV:7TC7:6DM7:Z24R:FOTL:FLQD:JTNI:OCYG:S66T:CNLN:FDAH
Debug mode (client): false
Debug mode (server): false
WARNING: No swap limit support
Experimental: true
```

```
Mar 25 11:39:02 dell kernel: [469502.427569] audit: type=1400 audit(1458895142.009:20495): apparmor="DENIED" operation="ptrace" profile="docker-default" pid=8351 comm="p
idof" requested_mask="trace" denied_mask="trace" peer="docker-default"
```
Removing AppArmod profile for docker and reloading service helped as suggested above.

**Comment by guss77** [comment]
This is still an issue with 1.11, after upgrading from an older release. The workaround for removing the old `/etc/apparmor.d/docker` configuration and restarting apparmor worker for me, but I really think that this should be handled by a `postinst` script in the `docker-engine` deb package.
